core/src/main/kotlin/com/expediagroup/sdk/core/model/paging/Paginator.kt

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (C) 2022 Expedia, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.expediagroup.sdk.core.model.paging
+
+import com.expediagroup.sdk.core.client.Client
+import com.expediagroup.sdk.core.model.Response
+import io.ktor.client.statement.HttpResponse
+
+sealed class BasePaginator<R, T>(
+    private val client: Client,
+    firstResponse: Response<T>,
+    private val getBody: suspend (HttpResponse) -> T
+) : Iterator<R> {
+    private var state: ResponseState<T> = DefaultResponseState(firstResponse)
+    val paginationTotalResults: Long = firstResponse.headers["pagination-total-results"]?.getOrNull(0)?.toLongOrNull() ?: 0
```

Comment:
Since the Paginator is no longer a mustache template, we can refactor string literals into a constant object